manta-trusted-setup/src/groth16/ceremony/server.rs

```diff
@@ -0,0 +1,164 @@
+        size: Array<StateSize, CIRCUIT_COUNT>,
+    ) -> Self {
+        let coordinator = Coordinator {
+            registry,
+            queue: Default::default(),
+            participant_lock: Default::default(),
+            state,
+            challenge,
+            latest_contributor: None,
+            latest_proof: None,
+            size,
+            round: 0,
+        };
+        Self {
+            coordinator: Arc::new(Mutex::new(coordinator)),
+            recovery_path,
+        }
+    }
+
+    /// Preprocess a request by checking nonce and verifying signature.
+    #[inline]
+    pub fn preprocess_request<T>(
+        registry: &mut R,
+        request: &Signed<T, C>,
+    ) -> Result<(), CeremonyError<C>>
+    where
+        T: Serialize,
+    {
+        if registry.has_contributed(&request.identifier) {
+            return Err(CeremonyError::AlreadyContributed);
+        }
+        let participant_nonce = registry
+            .get(&request.identifier)
+            .ok_or_else(|| CeremonyError::NotRegistered)?
+            .get_nonce();
+        if !check_nonce(&participant_nonce, &request.nonce) {
+            return Err(CeremonyError::NonceNotInSync(participant_nonce));
+        };
+        let mut participant = match registry.get(&request.identifier) {
+            Some(participant) => participant,
+            None => unreachable!("participant registration has been checked"),
+        };
+        verify::<T, C::SignatureScheme>(
+            participant.verifying_key(),
+            participant_nonce,
+            &request.message,
+            &request.signature,
+        )
+        .map_err(|_| CeremonyError::BadRequest)?;
+        participant.increment_nonce();
+        Ok(())
+    }
+
+    /// Gets the server state size and the current nonce of the participant.
+    #[inline]
+    pub async fn start(
+        self,
+        request: C::Identifier,
+    ) -> Result<(ServerSize<CIRCUIT_COUNT>, Nonce<C>), CeremonyError<C>> {
+        let coordinator = self
+            .coordinator
+            .lock()
+            .expect("acquiring a lock is not allowed to fail");
+        Ok((
+            coordinator.size.clone().into(),
+            coordinator
+                .nonce(&request)
+                .ok_or_else(|| CeremonyError::NotRegistered)?,
+        ))
+    }
+
+    /// Queries the server state
+    #[inline]
+    pub async fn query(
+        self,
+        request: Signed<QueryRequest, C>,
+    ) -> Result<QueryResponse<C, CIRCUIT_COUNT>, CeremonyError<C>>
+    where
+        C::Identifier: Serialize,
+        State<C::Pairing>: CanonicalSerialize + CanonicalDeserialize,
+        Challenge<C>: CanonicalSerialize + CanonicalDeserialize,
+    {
+        let mut coordinator = self.coordinator.lock();
+        Self::preprocess_request(&mut coordinator.registry, &request)?;
+        if !coordinator.is_in_queue(&request.identifier)? {
+            coordinator.enqueue_participant(&request.identifier)?;
```

Comment:
`enqueue_if_missing` in `manta-util`: if missing, enqueue. In both cases, return the position. Merge Line 147 ~150, 155.